### DISCUSSION TITLE
How to "reinitialize" an existing database instance?

### DISCUSSION THREAD

**Discussion by Adam-Langley** [simolus3/drift#2614] [Q&A] [answered]
My database implementation is injected into all the requisite classes. When the user logs out of the application, all local data is deleted, including the database.
The user may then log back in (i.e, effectively changing the logged in user) - so I need the database to be created again.
I cannot "reinject" a new database instance.
I see the internal connection inside the database class is final - is there a way to ask a DB to recreate itself?
I.e, how do I reinstate the database instance, its sqlite file, tables etc, without having to swap out the database instance in all the locations my application uses it?

I know I can use various patterns to achieve the same effect (locators, adapters, etc), however it would clearly be nice to just be able to reinitialize the existing instance.

Thanks,

**Comment by simolus3** [comment, marked as answer]
As you've seen, the "closed" state is also final internally in drift. You can sort of work around this by applying a workaround at a lower level, for instance on the connection interface between drift and the native database implementation.

With the class given below, you can inject a `ReconnectableExecutable` with the initial executor to construct the database. To close it, you can call `.close()` on the `ReconnectableExecutable` (not the database instance). Finally, you can use `reconnect` to make the database usable again.

```dart
import 'package:drift/drift.dart';

class ReconnectableExecutable extends QueryExecutor {
  QueryExecutor? _inner;

  ReconnectableExecutable(QueryExecutor this._inner);

  void reconnect(QueryExecutor newConnection) {
    _inner = newConnection;
  }

  QueryExecutor _resolve() {
    if (_inner case final QueryExecutor active) {
      return active;
    } else {
      throw StateError('This database is currently closed');
    }
  }

  @override
  TransactionExecutor beginTransaction() {
    return _resolve().beginTransaction();
  }

  @override
  SqlDialect get dialect => _resolve().dialect;

  @override
  Future<bool> ensureOpen(QueryExecutorUser user) {
    return _resolve().ensureOpen(user);
  }

  @override
  Future<void> runBatched(BatchedStatements statements) {
    return _resolve().runBatched(statements);
  }

  @override
  Future<void> runCustom(String statement, [List<Object?>? args]) {
    return _resolve().runCustom(statement, args);
  }

  @override
  Future<int> runDelete(String statement, List<Object?> args) {
    return _resolve().runDelete(statement, args);
  }

  @override
  Future<int> runInsert(String statement, List<Object?> args) {
    return _resolve().runInsert(statement, args);
  }

  @override
  Future<List<Map<String, Object?>>> runSelect(
      String statement, List<Object?> args) {
    return _resolve().runSelect(statement, args);
  }

  @override
  Future<int> runUpdate(String statement, List<Object?> args) {
    return _resolve().runUpdate(statement, args);
  }

  @override
  Future<void> close() async {
    await _resolve().close();
    _inner = null;
  }
}
```

**Reply by Adam-Langley** [reply]
Thank you @simolus3
